stoppable_workers.go

```diff
@@ -44,16 +45,49 @@ func NewBackgroundStoppableWorkers(workers ...func(context.Context)) *StoppableW
 	return sw
 }
 
-// Add starts up a goroutine for the passed-in function. Workers:
-//
-//   - MUST respond appropriately to errors on the context parameter.
-//   - MUST NOT add more workers to the `StoppableWorkers` group to which
-//     they belong.
+// NewStoppableWorkerWithTicker creates a `StoppableWorkers` object with a single worker that gets
+// called every `tickRate`. Calls to the input `worker` function are serialized. I.e: a slow "work"
+// iteration will just slow down when the next one is called.
+func NewStoppableWorkerWithTicker(tickRate time.Duration, workFn func(context.Context)) *StoppableWorkers {
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	sw := &StoppableWorkers{ctx: ctx, cancelFunc: cancelFunc}
+	sw.workers.Add(1)
+	PanicCapturingGo(func() {
+		defer sw.workers.Done()
+
+		timer := time.NewTicker(tickRate)
+		defer timer.Stop()
+		for {
+			select {
+			case <-ctx.Done():
```

Comment:
I normally prefer to _not_ see this extra "select" statement. But I left it in here because otherwise I'd have to justify why my test assertion needs to check for ">= 1" instead of exactly 1.

And in the process of justifying it I felt pretty petty for avoiding writing this code.

That said, I'd still prefer to not see these. Most use cases do not need them at all. But using tickers with a small timeout that kicks off jobs is one of the exceptions.

Reply:
I buy that this is an exception. I agree that this pattern is generally confusing and usually unnecessary, though.